### PR TITLE
Support ~j and ~J sigils.

### PR DESCRIPTION
**Pull request by azizk**
Hi!

I added a Sigil module because I thought it would be nice to have JSON sigils similar to `~y` in YamlElixir.

### CODE REVIEW

File: lib/sigil.ex [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  Handles the sigil `~j` for JSON strings.
+
+  Calls `Jason.decode!/2` with modifiers mapped to options.
+
+  ## Modifiers
+
+    * `a` - keys are converted to atoms using String.to_atom/1
+    * `A` - keys are converted to atoms using String.to_existing_atom/1
+    * `r` - when possible tries to create a sub-binary into the original
+    * `c` - always copies the strings
```
Comment (michalmuskala) [inline lib/sigil.ex:12]: Let's describe those in terms of equivalent options from `Jason.decode`

**Review by michalmuskala — COMMENTED** [review]
I really like this idea! Thank you for contributing those.

File: lib/sigil.ex (hunk 1 above)
Comment (azizk) [reply lib/sigil.ex]: Fixed: referring to Jason.decode/2 instead now.

**Comment by azizk** [comment]
Do you think the module name Sigil is fine? Or should it be Sigils?

**Comment by michalmuskala** [comment]
I think wither one is fine. Given the name in the other library is `Sigil`, let's use the same.

**Comment by michalmuskala** [comment]
Thank you! :heart:

**Comment by michalmuskala** [comment]
I pushed a commit https://github.com/elixir-lang/elixir/blob/31751f67da68470e655c884d64a8cbbd7f2e4903/lib/elixir/lib/kernel.ex#L5253-L5287 that converts the sigils to macros to decode the strings at compile-time when possible

**Comment by azizk** [comment]
Thanks for being so responsive and merging! :heart:

One thing I noticed is you added a sentence to the doc of `sigil_j`: "Given a string literal without interpolations, decodes the string at compile-time."
I think even with interpolations Elixir evaluates the decoding at compile-time. You don't even need macros for that. E.g.:
```
  x = &Jason.decode/2
  @y &Jason.decode!/2
  ~j'"#{inspect(x)}; #{inspect(@y)}"' |> IO.inspect(label: "~j with interpolations")
```
Or does it not happen in older Elixir versions without macros? Maybe I'm misunderstanding something... :)

**Comment by michalmuskala** [comment]
If you define the sigil as `def` it always happens at runtime. Only macros are expanded at compile-time. With `def sigil_j` this would be exactly the same as calling `Jason.decode!("\"#{inspect(x)}; #{inspect(@y)}\"")` except for the escapes & shorter syntax. With `defmacro`, it will be the same as writing a literal string.

**Pull request merged**